docs/tutorial/tests/steps/library_steps.py

```diff
@@ -1,34 +1,72 @@
-from helper_methods.library_catalog import Catalog
-from helper_methods.verification_helper_methods import verify_returned_books
-from pytest import fixture
-from pytest_bdd_ng import given, parsers, step, then, when
+import re
+from typing import Literal
 
+from src.catalog import Book, Catalog
 
-@fixture
-def context()
-    """ Create a placeholder object to use in place of Cucumber's context object. The context object allows us to pass state between steps. """
-    class dummy():
-        pass
+from messages import DataTable, Step  # type:ignore[attr-defined]
+from pytest_bdd import given, step, then, when
 
 
-    return dummy()
+def get_books_from_data_table(data_table: DataTable):
+    # Gherkin data-tables have no title row by default, but we could use if we want.
+    step_data_table_titles = [cell.value for cell in data_table.rows[0].cells]
+    assert step_data_table_titles == ["Author", "Title"]
 
+    author_and_title_list = [[cell.value for cell in row.cells] for row in data_table.rows[1:]]
 
-@given("these books in the catalog")
-def these_books_in_the_catalog(step):
-    context.catalog = Catalog()
-    context.catalog.add_books_to_catalog(step.data_table)
+    books = [Book(author, title) for author, title in author_and_title_list]
+    return books
 
 
-@when(parsers.re("a (?P<search_type>name|title) search is performed for " +
-"(?P<search_term>.+)"))
-def a_SEARCH_TYPE_is_performed_for_SEARCH_TERM(search_type, search_term):
+# Steps to be used in scenarios are defined with special decorators
+@given(
+    "these books in the catalog",
+    # Steps are allowed to inject new fixtures or overwrite existing ones
+    target_fixture="catalog",
+)
+def these_books_in_the_catalog(
+    # `step` fixture is injected by pytest DI mechanism into scope of step by default
+    step: Step,
+):
```

Comment:
I think this is better (easier to understand) than the way I wrote it, particularly with the explanations of fixture injection; I will rewrite my tutorial to look more like this. I would spell out 'DI'; not everyone will know that it means 'Dependency Injection'.